core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,717 @@
+}
+
+static struct mmu_pgt *core_mmu_pgt_alloc(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = NULL;
+
+	if (prtn->pgts_used >= RISCV_MMU_MAX_PGTS) {
+		debug_print("%u pgts exhausted", RISCV_MMU_MAX_PGTS);
+		panic();
+		return NULL;
+	}
+
+	pgt = &prtn->pool_pgts[prtn->pgts_used++];
+
+	memset(pgt, 0, RISCV_MMU_PGT_SIZE);
+
+	debug_print("pgts used %u / %u", prtn->pgts_used, RISCV_MMU_MAX_PGTS);
+
+	return pgt;
+}
+
+static void core_init_mmu_prtn_ta_core(struct mmu_partition *prtn __unused,
+				       unsigned int core __unused)
+{
+	/*
+	 * user_va_idx is the index in CORE_MMU_BASE_TABLE_LEVEL.
+	 * The entry holds pointer to the user mapping table in next level
+	 * that changes per core. Therefore, nothing to do.
+	 */
+}
+
+static void core_init_mmu_prtn_ta(struct mmu_partition *prtn)
+{
+	unsigned int core = 0;
+
+	assert(user_va_idx != -1);
+
+	memset(prtn->user_pgts, 0, CFG_NUM_THREADS * RISCV_MMU_PGT_SIZE);
+	for (core = 0; core < CFG_TEE_CORE_NB_CORE; core++)
+		core_init_mmu_prtn_ta_core(prtn, core);
+}
+
+static void core_init_mmu_prtn_tee(struct mmu_partition *prtn,
+				   struct tee_mmap_region *mm)
+{
+	size_t n = 0;
+	void *pgt = core_mmu_get_root_pgt_va(prtn);
+
+	memset(pgt, 0, RISCV_MMU_PGT_SIZE);
+	memset(prtn->pool_pgts, 0, RISCV_MMU_MAX_PGTS * RISCV_MMU_PGT_SIZE);
+
+	for (n = 0; !core_mmap_is_end_of_table(mm + n); n++)
+		if (!core_mmu_is_dynamic_vaspace(mm + n))
+			core_mmu_map_region(prtn, mm + n);
+}
+
+void tlbi_mva_range(vaddr_t va, size_t len,
+		    size_t granule)
+{
+	assert(granule == CORE_MMU_PGDIR_SIZE || granule == SMALL_PAGE_SIZE);
+	assert(!(va & (granule - 1)) && !(len & (granule - 1)));
+
+	while (len) {
+		tlbi_mva_allasid_nosync(va);
+		len -= granule;
+		va += granule;
+	}
+}
+
+void tlbi_mva_range_asid(vaddr_t va, size_t len,
+			 size_t granule, uint32_t asid)
+{
+	assert(granule == CORE_MMU_PGDIR_SIZE || granule == SMALL_PAGE_SIZE);
+	assert(!(va & (granule - 1)) && !(len & (granule - 1)));
+
+	while (len) {
+		tlbi_mva_asid_nosync(va, asid);
+		len -= granule;
+		va += granule;
+	}
+}
+
+TEE_Result cache_op_inner(enum cache_op op, void *va, size_t len)
+{
+	switch (op) {
+	case DCACHE_CLEAN:
+		dcache_op_all(DCACHE_OP_CLEAN);
+		break;
+	case DCACHE_AREA_CLEAN:
+		dcache_clean_range(va, len);
+		break;
+	case DCACHE_INVALIDATE:
+		dcache_op_all(DCACHE_OP_INV);
+		break;
+	case DCACHE_AREA_INVALIDATE:
+		dcache_inv_range(va, len);
+		break;
+	case ICACHE_INVALIDATE:
+		icache_inv_all();
+		break;
+	case ICACHE_AREA_INVALIDATE:
+		icache_inv_range(va, len);
+		break;
+	case DCACHE_CLEAN_INV:
+		dcache_op_all(DCACHE_OP_CLEAN_INV);
+		break;
+	case DCACHE_AREA_CLEAN_INV:
+		dcache_cleaninv_range(va, len);
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+	return TEE_SUCCESS;
+}
+
+unsigned int asid_alloc(void)
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+	unsigned int r = 0;
+	int i = 0;
+
+	bit_ffc(&g_asid, BIT(g_asid), &i);
```

Comment:
Compiler generates the following warning due to different integer type:

```bash
  CC      out/riscv/core/arch/riscv/mm/core_mmu_arch.o
In file included from core/arch/riscv/mm/core_mmu_arch.c:7:
core/arch/riscv/mm/core_mmu_arch.c: In function 'asid_alloc':
lib/libutils/ext/include/bitstring.h:123:20: warning: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Wsign-compare]
  123 |         if (_value >= nbits) \
      |                    ^~
core/arch/riscv/mm/core_mmu_arch.c:385:9: note: in expansion of macro 'bit_ffc'
  385 |         bit_ffc(&g_asid, BIT(g_asid), &i);
      |         ^~~~~~~
```

Is it possible to replace `BIT(g_asid)` with constant ?

Reply:
`g_asid` is not constant ?

Reply:
It is because `BIT()` macro will be expanded  to unsigned type.
The warning will gone if we replace `BIT(g_asid)` with signed type.